50 add foreign key restraing to admin id on neighborhoods model

radu-constantin opened the pull request.

RobertRodes reviewed (commented):
I don't understand why there's an `adminOf` field in the `User` model. Similarly, I don't understand why there's a `neighborhoods_users` field in the `User` model. The `users` entity is the one side of a one-to-many relationship in both cases, so the foreign key relation should be defined in the many table.

Ok, after reading up on this in the Prisma doc (https://www.prisma.io/docs/concepts/components/prisma-schema/relations), I have a better idea what you're up to. I cringe, though, at defining a property that's going to be empty most of the time with the `User.adminOf` property.

Semantically, if we're being consistent with the example, your `Neighborhood.NeighborhoodUsers` should just be `Neighborhood.Users`, and your `User.NeighborhoodUsers` should just be `User.Neighborhoods`. Note the plural on both; that represents the one-to-many relationship semantically.

In any case, the way you have it now, I'm not clear that you're actually defining the relationship, since there's no `@relation` definition. Based on the doc for implicit many-to-many relations, we need to do it like this:

1. change line 78/80 to `Neighborhoods` @relation("neighborhoods")`
2. change line 24/25 to `Users` @relation("users")`

I think this will pick up the implicit many-to-many relationship. 

Another concern is what to do with `onUpdate`. Perhaps we should consider cascading that as well. Not that it makes a heck of a lot of difference: what it does is if you change a primary key specifically (and who does that?), all tables referencing it as a foreign key get changed as well.

radu-constantin commented:
It's actually an explicit many to many relationship right now. I didn't try to implement an implicit one. I could try and do it though.
When I converted your schema to Prisma, it automatically created an _explicit_ many to many relationship.

RobertRodes commented:
So, you're saying that this is the way that Prisma handled the mapping? If so, it doesn't look consistent with the documentation for explicit m-n relations. 

According to this: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations, the explicit relation has these characteristics:

-     Two models that have a many-to-many relation, such as Category and Post
-     One model that represents the relation table, such as CategoriesOnPosts (also sometimes called JOIN, link or pivot table) in the underlying database

This suggests that we would have a separate model called `NeighborhoodsOnUsers` that looks something like this:
```
model NeighborhoodsOnUsers {
  neighborhood       Neighborhood     @relation(fields: [neighborhoodId], references: [id])
  neighborhoodId     Int // relation scalar field (used in the `@relation` attribute above)
  user  User @relation(fields: [userId], references: [id])
  userId Int // relation scalar field (used in the `@relation` attribute above)
  @@id([neighborhoodId, userId])
}
```
I've just transliterated the example they give with `Categories` and `Posts`. Since we're not keeping any data but the composite key, ours is a good candidate for an implicit relation.

radu-constantin commented:
I might be misunderstanding the issue, but we have such a model. It's in the prisma.schema
```
model NeighborhoodUsers {
  neighborhood_id Int
  user_id         Int
  active          Boolean?
  neighborhoods   Neighborhood @relation(fields: [neighborhood_id], references: [id], onDelete: Cascade, onUpdate: NoAction)
  users           User         @relation(fields: [user_id], references: [id], onDelete: Cascade, onUpdate: NoAction)
  @@id([neighborhood_id, user_id])
  @@map("neighborhoods_users")
}
```
I agree, we could simply use an implicit M-M relationship. The only problem is that we would have to loose the `active` field. I don't see a big issue with that. We don't really need to track active users.

RobertRodes commented:
I forgot about the `active` field. And I'm the one who was misunderstanding. I've been looking at an incomplete version of the schema! Ok, that looks good. At least I'm picking up how Prisma models relational DBs. :)

Let me see if I can figure out where to see the schema. If I don't find it, I'll ask.

RobertRodes commented:
Ok, found it. Another way would be to pull it into another `admins` table, with a neighborhood id and a user id.

radu-constantin commented:
Ok, so like an explitic many-to-many relationship for admins.
By the way, I've managed to transform Neighborhood - Users relationship to an implicit M-M. I think it's better this way, and it will allow for simpler Prisma queries.

RobertRodes commented:
No, it would still be one to many. Only one admin for each neighborhood, possibly many neighborhoods with the same admin.

You would just set it up with the neighborhood id as the primary key, and then add a user id for each neighborhood id. That user id could be the same one for different neighborhoods.

RobertRodes commented:
I thought I found this but I haven't. I looked at schema.prisma, but that's on main, so it doesn't have your changes. Where can I see schema.prisma with your changes, but without all the red and green stuff? Just the file as it stands now?

radu-constantin commented:
Correct, one to many!

radu-constantin commented:
I'm not sure if having that 'adminOf' property is such a big drawback. It's a bit of an eye sore, but you can check easily if a user is an admin without having to make additional queeies.

radu-constantin commented:
You can either download the branch on your local machine (I can explain how if you want) or if you want to view directly on Github do it like this. 
1. Click the commits tab at the top of this page
![Commits](https://user-images.githubusercontent.com/40914437/230971929-71c9eceb-8f58-40d7-b8a6-70c39734b2e8.jpg)
2. Click this button that looks like an arrow tag.
![Commits](https://user-images.githubusercontent.com/40914437/230972259-7adf13fd-8d02-4ab5-b179-b08cc9fefbc0.jpg)

You will able to see the repository as I see it in my branch.

nina-dk commented:
I'm getting an error when I run `npm run seed` and the tests fail (`npm run test`). I've run `npm run reinitialize-db`.

```
TSError: ⨯ Unable to compile TypeScript:
prisma/seed.ts:80:16 - error TS2551: Property 'neighborhoodUsers' does not exist on type 'PrismaClient<{ log: "query"[]; }, never, false>'. Did you mean 'neighborhood'?

80   await prisma.neighborhoodUsers.createMany({
                  ~~~~~~~~~~~~~~~~~

  node_modules/.prisma/client/index.d.ts:214:7
    214   get neighborhood(): Prisma.NeighborhoodDelegate<GlobalReject>;
              ~~~~~~~~~~~~
    'neighborhood' is declared here.
```

RobertRodes commented:
Possible version problem? We’ve been tinkering with the schema quite a bit last couple of days.

> On Apr 10, 2023, at 2:24 PM, Antonina Kallinteri ***@***.***> wrote:
> 
> 
> I'm getting an error when I run npm run seed and the tests fail (npm run test). I've run npm run reinitialize-db.
> 
> TSError: ⨯ Unable to compile TypeScript:
> prisma/seed.ts:80:16 - error TS2551: Property 'neighborhoodUsers' does not exist on type 'PrismaClient<{ log: "query"[]; }, never, false>'. Did you mean 'neighborhood'?
> 
> 80   await prisma.neighborhoodUsers.createMany({
>                   ~~~~~~~~~~~~~~~~~
> 
>   node_modules/.prisma/client/index.d.ts:214:7
>     214   get neighborhood(): Prisma.NeighborhoodDelegate<GlobalReject>;
>               ~~~~~~~~~~~~
>     'neighborhood' is declared here.
> —
> Reply to this email directly, view it on GitHub <https://github.com/neighborhood-app/backend/pull/51#issuecomment-1502213535>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJBIA6W5SDKMHLYOR7FYLNDXARNARANCNFSM6AAAAAAWY5CH3Y>.
> You are receiving this because your review was requested.
>

radu-constantin commented:
Will fix it tomorrow. It's because I've changed the Neighborhood Users relationship from **explicit** to implicit many-to-many.

RobertRodes commented:
Yup. Version problem. :)

radu-constantin commented:
Fixed!

nina-dk commented:
I think there's some references to the old table in `tests/seed.ts` as well that cause the tests to fail.

radu-constantin commented:
Fixed! Updated the test seed file. 
Another problem is that jest is currently also running the tests in the build folder (we don't need to test those). I think @MaxHass12 solved this in the branch he's working on.

radu-constantin commented:
Yes, without the build folder everything should work fine.

nina-dk commented on prisma/schema.prisma:
What does adding the `admin` row do exactly? Is that how `admin_id` is set as a foreign key?

MaxHass12 reviewed (approved):
Everything looks good.


Approved !
